src/lib/validationRules/index.ts

```diff
@@ -0,0 +1,162 @@
+import validApplicationType from "./validApplicationType/validApplicationType";
+import rightAuthenticationMethod from "./rightAuthenticationMethod/rightAuthenticationMethod";
+import noUnknownFields from "./noUnknownFields/noUnknownFields";
+import validScope from "./validScope/validScope";
+import validGrantTypes from "./validGrantTypes/validGrantTypes";
+import noUnsetClientUri from "./noUnsetClientUri/noUnsetClientUri";
+
+import { ValidationResults, ValidationRule } from "../types";
+
+export const offlineRules: ValidationRule[] = [
+  decentClientName,
+  validUriFields,
+  validContext,
+  validRedirectUris,
+  refreshTokenRule,
+  validGrantTypes,
+  validScope,
+  noUnknownFields,
+  defaultMaxAgeRule,
+  validApplicationType,
+  rightAuthenticationMethod,
+  validResponseType,
+  sameDomainForRedirectUris,
+  noLocalAndRemoteRedirectUris,
+  noUnsetClientUri,
+  // We omit localization validation as it is not supported by ESS or NSS (as of 2022-08-17) but leave it here for spec-compliance guidance.
+];
+
+export async function validateDocument(
+  jsonDocument: string,
+  rules: ValidationRule[]
+): Promise<ValidationResults> {
+  // try parsing first..
+  let clientIdDocument = {};
+  try {
+    clientIdDocument = JSON.parse(jsonDocument);
+  } catch {
+    return [
+      {
+        rule: {
+          type: "local",
+          name: "Document must be valid JSON",
+          description: "The document must be a valid JSON string.",
+        },
+        status: "error",
+        title: "Invalid JSON",
+        description: "The document could not be parsed to JSON.",
+        affectedFields: [],
+      },
+    ];
+  }
+
+  const validationPromises = rules.map(async (rule) => {
+    const results = await rule.check({ document: clientIdDocument });
+    return results.map((result) => ({
+      rule: rule.rule,
+      ...result,
+    }));
+  });
+
+  return (await Promise.all(validationPromises)).flat();
+}
+
+export async function requestRemoteValidation(
+  documentIri: string
+): Promise<ValidationResults> {
+  let response: Response;
+  try {
+    response = await await fetch(
+      `/api/validate-remote-document?documentIri=${documentIri}`,
+      {
+        method: "POST",
+      }
+    );
+  } catch (error) {
+    return [
+      {
+        status: "error",
+        affectedFields: [],
+        rule: { description: "", name: "", type: "remote" },
+        title: `Fetching resource was not successful`,
+        description: `An error ocured while trying to fetch the resource: ${error}`,
+      },
+    ];
+  }
+  let jsonResponse: any;
+  try {
+    jsonResponse = await response.json();
+  } catch (error) {
+    return [
+      {
+        status: "error",
+        affectedFields: [],
+        rule: { description: "", name: "", type: "remote" },
+        title: `Fetching resource was not successful`,
+        description: `The validation result could not be parsed.`,
+      },
+    ];
+  }
+
+  if (jsonResponse?.error) {
```

Comment:
You could also make the `validateRemoteDocument` API just always return `document.results` here, which conforms to the ValidationResult type. So you don't have `document.error` vs `document.results` — there is only results.